reflex-dom-core/src/Reflex/Dom/Modals/Base.hs

```diff
@@ -0,0 +1,181 @@
+  type DomBuilderSpace (ModalsT t m) = DomBuilderSpace m
+  element = coerce (element :: Text -> ElementConfig er t (DomBuilderSpace m) -> Unwrapped (ModalsT t m a) -> Unwrapped (ModalsT t m (Element er (DomBuilderSpace m) t, a)))
+  inputElement = coerce (inputElement :: InputElementConfig er t (DomBuilderSpace m) -> Unwrapped (ModalsT t m (InputElement er (DomBuilderSpace m) t)))
+  textAreaElement = coerce (textAreaElement :: TextAreaElementConfig er t (DomBuilderSpace m) -> Unwrapped (ModalsT t m (TextAreaElement er (DomBuilderSpace m) t)))
+  selectElement = coerce (selectElement :: SelectElementConfig er t (DomBuilderSpace m) -> Unwrapped (ModalsT t m a) -> Unwrapped (ModalsT t m (SelectElement er (DomBuilderSpace m) t, a)))
+  wrapRawElement e = (view _Unwrapped :: forall a. Unwrapped (ModalsT t m a) -> ModalsT t m a) . wrapRawElement e
+
+instance HasDocument m => HasDocument (ModalsT t m)
+instance HasJSContext m => HasJSContext (ModalsT t m) where
+  type JSContextPhantom (ModalsT t m) = JSContextPhantom m
+  askJSContext = ModalsT askJSContext
+#ifndef ghcjs_HOST_OS
+instance MonadJSM m => MonadJSM (ModalsT t m)
+#endif
+
+withModalLayerClass :: forall t m a. (Reflex t, MonadFix m, DomBuilder t m, MonadHold t m)
+                    => Text
+                    -> ModalsT t m a
+                    -> m a
+withModalLayerClass c (ModalsT a) = elAttr "div" ("class" =: c) $ do
+  rec (result, requests) <- do
+        runRequesterT a modalDone
+      modalDone <- modalInner requests
+  return result
+
+modalInner :: forall t m k. (GCompare k, MonadHold t m, MonadFix m, DomBuilder t m)
+           => Event t (DMap k (Compose (ModalsT t m) (Event t)))
+           -> m (Event t (DMap k Maybe))
+modalInner rqs' = do
+  rec let rqs = fmap toInsertionsDMap rqs'
+          dones = fmap toDeletionsDMap modalDone
+      (m0, mpatch) <- traverseDMapWithKeyWithAdjust (\k v -> Compose <$> modalBody k v) DMap.empty (rqs <> dones)
+      modal <- fmap (merge . mapKeyValuePairsMonotonic wrapArgs) . incrementalToDynamic <$> holdIncremental m0 mpatch
+      let modalDone = fmap (mapKeyValuePairsMonotonic unwrapArgs) $ switch $ current modal
+  return modalDone
+ where
+  wrapArgs (k :=> v) = WrapArg k :=> getCompose v
+  unwrapArgs :: DSum (WrapArg Maybe k) Identity -> DSum k Maybe
+  unwrapArgs (WrapArg k :=> Identity v) = k :=> v
+  toInsertionsDMap :: DMap k v -> PatchDMap k v
+  toInsertionsDMap = coerce . DMap.map (ComposeMaybe . Just)
+  toDeletionsDMap :: forall k' v v'. DMap k' v -> PatchDMap k' v'
+  toDeletionsDMap = coerce . DMap.map (\(_ :: v a) -> ComposeMaybe (Nothing :: Maybe (v' a)))
+
+modalBody :: forall t m k a. (MonadHold t m, MonadFix m, DomBuilder t m)
+          => k a
+          -> Compose (ModalsT t m) (Event t) a
+          -> m (Event t (Maybe a))
+modalBody _ v = do
+  let style = T.intercalate ";"
+        [ "position:absolute"
+        , "left:0", "right:0", "top:0", "bottom:0"
+        , "background-color:rgba(0,0,0,0.1)"
```

Comment:
Giving the modal the ability to change the background sounds great.  The plan I've thought of for allowing that would be to parameterize ModalsT on the request functor type, so that instead of modals being specifically m (Event t a), they are f a, and the f a is interpreted by a function passed to withModalLayer.  I would probably also see if we can parameterize the response functor at the same time; I don't have a specific use case in mind yet, but I suspect they exist.